In [1]:
import wget

In [5]:
import pandas as pd

In [20]:
import numpy as np

In [2]:
data2 = 'https://raw.githubusercontent.com/alexeygrigorev/datasets/master/AB_NYC_2019.csv'

In [3]:
ny_airbnb = wget.download(data2)

100% [......................................................] 7077973 / 7077973

In [6]:
ny_abnb = pd.read_csv(ny_airbnb)

In [7]:
ny_abnb.head()

id                                              name  host_id  \
0  2539                Clean & quiet apt home by the park     2787   
1  2595                             Skylit Midtown Castle     2845   
2  3647               THE VILLAGE OF HARLEM....NEW YORK !     4632   
3  3831                   Cozy Entire Floor of Brownstone     4869   
4  5022  Entire Apt: Spacious Studio/Loft by central park     7192   

     host_name neighbourhood_group neighbourhood  latitude  longitude  \
0         John            Brooklyn    Kensington  40.64749  -73.97237   
1     Jennifer           Manhattan       Midtown  40.75362  -73.98377   
2    Elisabeth           Manhattan        Harlem  40.80902  -73.94190   
3  LisaRoxanne            Brooklyn  Clinton Hill  40.68514  -73.95976   
4        Laura           Manhattan   East Harlem  40.79851  -73.94399   

         room_type  price  minimum_nights  number_of_reviews last_review  \
0     Private room    149               1                  9  2018-10-19   
1  Entire home/apt    225               1                 45  2019-05-21   
2     Private room    150               3                  0         NaN   
3  Entire home/apt     89               1                270  2019-07-05   
4  Entire home/apt     80              10                  9  2018-11-19   

   reviews_per_month  calculated_host_listings_count  availability_365  
0               0.21                               6               365  
1               0.38                               2               355  
2                NaN                               1               365  
3               4.64                               1               194  
4               0.10                               1                 0

In [8]:
ny_abnb.columns.str.lower

<bound method StringMethods.lower of <pandas.core.strings.accessor.StringMethods object at 0x116458ac0>>

In [9]:
ny_abnb.columns = ny_abnb.columns.str.replace(' ','_')

In [10]:
cols = ["neighbourhood_group", "room_type", "latitude", 'longitude','price','minimum_nights','number_of_reviews','reviews_per_month','calculated_host_listings_count','availability_365']

In [16]:
ny_abnb_sub = ny_abnb[cols]

In [17]:
ny_abnb_sub.isnull().sum()

neighbourhood_group                   0
room_type                             0
latitude                              0
longitude                             0
price                                 0
minimum_nights                        0
number_of_reviews                     0
reviews_per_month                 10052
calculated_host_listings_count        0
availability_365                      0
dtype: int64

In [21]:
ny_abnb_sub_miss = ny_abnb_sub.replace(np.nan, 0)

In [22]:
ny_abnb_sub_miss.isnull().sum()

neighbourhood_group               0
room_type                         0
latitude                          0
longitude                         0
price                             0
minimum_nights                    0
number_of_reviews                 0
reviews_per_month                 0
calculated_host_listings_count    0
availability_365                  0
dtype: int64

In [23]:
ny_abnb_sub_miss.head()

neighbourhood_group        room_type  latitude  longitude  price  \
0            Brooklyn     Private room  40.64749  -73.97237    149   
1           Manhattan  Entire home/apt  40.75362  -73.98377    225   
2           Manhattan     Private room  40.80902  -73.94190    150   
3            Brooklyn  Entire home/apt  40.68514  -73.95976     89   
4           Manhattan  Entire home/apt  40.79851  -73.94399     80   

   minimum_nights  number_of_reviews  reviews_per_month  \
0               1                  9               0.21   
1               1                 45               0.38   
2               3                  0               0.00   
3               1                270               4.64   
4              10                  9               0.10   

   calculated_host_listings_count  availability_365  
0                               6               365  
1                               2               355  
2                               1               365  
3                               1               194  
4                               1                 0

In [26]:
ny_abnb_sub_miss.dtypes 

neighbourhood_group                object
room_type                          object
latitude                          float64
longitude                         float64
price                               int64
minimum_nights                      int64
number_of_reviews                   int64
reviews_per_month                 float64
calculated_host_listings_count      int64
availability_365                    int64
dtype: object

In [24]:
for col in ny_abnb_sub_miss.columns:
    print(col)
    print(ny_abnb_sub[col].unique()[:5])
    print(ny_abnb_sub[col].nunique())
    print()

neighbourhood_group
['Brooklyn' 'Manhattan' 'Queens' 'Staten Island' 'Bronx']
5

room_type
['Private room' 'Entire home/apt' 'Shared room']
3

latitude
[40.64749 40.75362 40.80902 40.68514 40.79851]
19048

longitude
[-73.97237 -73.98377 -73.9419  -73.95976 -73.94399]
14718

price
[149 225 150  89  80]
674

minimum_nights
[ 1  3 10 45  2]
109

number_of_reviews
[  9  45   0 270  74]
394

reviews_per_month
[0.21 0.38 0.   4.64 0.1 ]
938

calculated_host_listings_count
[6 2 1 4 3]
47

availability_365
[365 355 194   0 129]
366



In [27]:
from sklearn.model_selection import train_test_split

In [30]:
df_full_train , df_test = train_test_split(ny_abnb_sub_miss, test_size=0.2, random_state=42)

In [32]:
df_train , df_val = train_test_split(df_full_train, test_size=0.25, random_state=42)

In [33]:
len(df_full_train), len(df_test), len(df_train), len(df_val)

(39116, 9779, 29337, 9779)

In [34]:
ny_abnb_sub_miss.neighbourhood_group.value_counts()

Manhattan        21661
Brooklyn         20104
Queens            5666
Bronx             1091
Staten Island      373
Name: neighbourhood_group, dtype: int64

## Question 1

### What is the most frequent observation (mode) for the column 'neighbourhood_group'?
### Solution : The column 'neighbourhood_group' has only 5 values namely ['Brooklyn' 'Manhattan' 'Queens' 'Staten Island' 'Bronx'] of which Manhattan is most frequent observation. 

In [35]:
corr_matrix = ny_abnb_sub_miss.corr()
print(corr_matrix)

                                latitude  longitude     price  minimum_nights  \
latitude                        1.000000   0.084788  0.033939        0.024869   
longitude                       0.084788   1.000000 -0.150019       -0.062747   
price                           0.033939  -0.150019  1.000000        0.042799   
minimum_nights                  0.024869  -0.062747  0.042799        1.000000   
number_of_reviews              -0.015389   0.059094 -0.047954       -0.080116   
reviews_per_month              -0.018758   0.138516 -0.050564       -0.124905   
calculated_host_listings_count  0.019517  -0.114713  0.057472        0.127960   
availability_365               -0.010983   0.082731  0.081829        0.144303   

                                number_of_reviews  reviews_per_month  \
latitude                                -0.015389          -0.018758   
longitude                                0.059094           0.138516   
price                                   -0.047954     

## Question 2

###     What are the two features that have the biggest correlation in this dataset?


### Solution : 
calculated_host_listings_count and availability_365 = 0.2257

number_of_reviews and availability_365 = 0.172

number_of_reviews and reviews_per_month = 0.589

minimum_nights and calculated_host_listings_count = 0.127 

### Therefore, number_of_reviews and reviews_per_month is the answer. 

In [38]:
ny_abnb_sub_miss['above_average'] = np.where(ny_abnb_sub_miss['price']>= 152, 1, 0)

In [39]:
ny_abnb_sub_miss.head()

neighbourhood_group        room_type  latitude  longitude  price  \
0            Brooklyn     Private room  40.64749  -73.97237    149   
1           Manhattan  Entire home/apt  40.75362  -73.98377    225   
2           Manhattan     Private room  40.80902  -73.94190    150   
3            Brooklyn  Entire home/apt  40.68514  -73.95976     89   
4           Manhattan  Entire home/apt  40.79851  -73.94399     80   

   minimum_nights  number_of_reviews  reviews_per_month  \
0               1                  9               0.21   
1               1                 45               0.38   
2               3                  0               0.00   
3               1                270               4.64   
4              10                  9               0.10   

   calculated_host_listings_count  availability_365  above_average  
0                               6               365              0  
1                               2               355              1  
2                               1               365              0  
3                               1               194              0  
4                               1                 0              0

In [40]:
from sklearn.model_selection import train_test_split

In [45]:
df_full_train , df_test = train_test_split(ny_abnb_sub_miss, test_size=0.2, random_state=42)

In [46]:
df_train , df_val = train_test_split(df_full_train, test_size=0.25, random_state=42)

In [112]:
len(df_full_train), len(df_test), len(df_train), len(df_val)

(39116, 9779, 29337, 9779)

In [48]:
df_full_train.head()

neighbourhood_group        room_type  latitude  longitude  price  \
32645            Brooklyn  Entire home/apt  40.71577  -73.95530    295   
23615           Manhattan     Private room  40.84917  -73.94048     70   
31183            Brooklyn     Private room  40.68993  -73.95947     58   
29260            Brooklyn  Entire home/apt  40.68427  -73.93118     75   
7275               Queens     Private room  40.74705  -73.89564     38   

       minimum_nights  number_of_reviews  reviews_per_month  \
32645               3                 11               0.87   
23615               2                  2               0.16   
31183               2                  0               0.00   
29260               3                 87               4.91   
7275                5                 13               0.25   

       calculated_host_listings_count  availability_365  above_average  
32645                               1                 1              1  
23615                               1                 0              0  
31183                               2                 0              0  
29260                               1               267              0  
7275                                1                 0              0

In [104]:
from sklearn.metrics import mutual_info_score 

In [105]:
mutual_info_score(df_full_train.above_average, df_full_train.neighbourhood_group)

0.0462226506346477

In [106]:
mutual_info_score(df_full_train.above_average, df_full_train.room_type)

0.1423898076642953

In [89]:
def mutual_info_score_cat_fac(s):
    return mutual_info_score(s, df_full_train.above_average)

In [86]:
categorical = ['neighbourhood_group' , 'room_type']

In [83]:
df_full_train[categorical].nunique()

neighbourhood_group    5
room_type              3
dtype: int64

In [107]:
def mutual_info_churn_score(series):
    return mutual_info_score(series, df_full_train.above_average)

In [108]:
mi = df_full_train[categorical].apply(mutual_info_churn_score)

In [109]:
mi.sort_values(ascending=False)

room_type              0.142390
neighbourhood_group    0.046223
dtype: float64

## Question 3

###  Which of these two variables has bigger score? Mututal Information 


### Solution : room_type has bigger mutual information score 
neighbourhood_group = 0.0462 
room_type = 0.142 

In [110]:
numerical = ['latitude' , 'longitude', 'minimum_nights','number_of_reviews', 'reviews_per_month', 'calculated_host_listings_count', 'availability_365']

In [111]:
from sklearn.feature_extraction import DictVectorizer

In [113]:
dv = DictVectorizer(sparse=False)

train_dict = df_train[categorical + numerical].to_dict(orient='records')
X_train = dv.fit_transform(train_dict)

val_dict = df_val[categorical + numerical].to_dict(orient='records')
X_val = dv.transform(val_dict)

In [117]:
df_train.shape

(29337, 11)

In [115]:
X_train[0]

array([ 50.     ,  13.     ,  40.7276 , -73.94495,   3.     ,   0.     ,
         1.     ,   0.     ,   0.     ,   0.     ,  29.     ,   0.7    ,
         1.     ,   0.     ,   0.     ])

In [116]:
X_train.shape

(29337, 15)

## We have 15 feature since the one-hot encoding converts the categorical variable into 5 + 3 i.e. 7 features since room_type has 3 unique values, neigh_type has 5 unique values making it total 7 features and 8 features from numerical features. 

In [118]:
y_train = df_train.above_average.values
y_val = df_val.above_average.values
y_test = df_test.above_average.values

In [119]:
y_train

array([0, 0, 0, ..., 1, 0, 0])

In [121]:
from sklearn.linear_model import LogisticRegression 

In [127]:
model = LogisticRegression(solver='lbfgs', C=1.0, random_state=42)
model.fit(X_train, y_train)

/Users/rashmi/mlzoomenv/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

In [128]:
model.coef_[0]

array([ 0.00292689,  0.00399591, -0.22707994, -0.0944945 , -0.01240595,
       -0.39587698,  0.07998583,  1.23969318, -0.79564353, -0.13077602,
       -0.00326021, -0.04113116,  1.63044242, -1.16284527, -0.47021467])

In [129]:
y_pred = model.predict_proba(X_val)[:,1]

In [130]:
y_pred

array([0.03271623, 0.56933217, 0.3966939 , ..., 0.09890387, 0.03405207,
       0.6395589 ])

In [131]:
final_decision = (y_pred >= 0.5)

In [132]:
final_decision.astype(int)

array([0, 1, 0, ..., 0, 0, 1])

In [133]:
y_val

array([0, 0, 1, ..., 0, 0, 0])

In [189]:
acc_main = (y_val == final_decision).mean().round(2)

In [190]:
acc_main

0.79

## Question 4 
### Calculate the accuracy on the validation dataset and rount it to 2 decimal digits.
### Accuracy is 0.79 rounded to 2 decimals 

In [136]:
neighbour_group = ['latitude' , 'longitude', 'minimum_nights','number_of_reviews', 'reviews_per_month', 'calculated_host_listings_count', 'availability_365', 'room_type']

In [137]:
room_type = ['latitude' , 'longitude', 'minimum_nights','number_of_reviews', 'reviews_per_month', 'calculated_host_listings_count', 'availability_365', 'neighbourhood_group']

In [162]:
number_of_reviews = ['latitude' , 'longitude', 'minimum_nights', 'reviews_per_month', 'calculated_host_listings_count', 'availability_365', 'room_type', 'neighbourhood_group']

In [139]:
reviews_per_month = ['latitude' , 'longitude', 'minimum_nights', 'number_of_reviews', 'calculated_host_listings_count', 'availability_365', 'room_type', 'neighbourhood_group']

In [144]:
len(neighbour_group), len(room_type), len(number_of_reviews), len(reviews_per_month)

(8, 8, 8, 8)

In [147]:
df_train[neighbour_group].iloc[:1].to_dict(orient='records')

[{'latitude': 40.7276,
  'longitude': -73.94495,
  'minimum_nights': 3,
  'number_of_reviews': 29,
  'reviews_per_month': 0.7,
  'calculated_host_listings_count': 13,
  'availability_365': 50,
  'room_type': 'Entire home/apt'}]

In [148]:
dicts_train_ng = df_train[neighbour_group].to_dict(orient='records')
dicts_val_ng = df_val[neighbour_group].to_dict(orient='records')

In [149]:
dicts_train_rt = df_train[room_type].to_dict(orient='records')
dicts_val_rt = df_val[room_type].to_dict(orient='records')

In [163]:
dicts_train_nor = df_train[number_of_reviews].to_dict(orient='records')
dicts_val_nor = df_val[number_of_reviews].to_dict(orient='records')

In [151]:
dicts_train_rpm = df_train[reviews_per_month].to_dict(orient='records')
dicts_val_rpm = df_val[reviews_per_month].to_dict(orient='records')

In [ ]:
X_val_ng = dv_ng.transform(dicts_val_ng)
X_val_rt = dv_rt.transform(dicts_val_rt)
X_val_nor = dv_nor.transform(dicts_val_nor)
X_val_rpm = dv_rpm.transform(dicts_val_rpm)

In [154]:
dv_ng = DictVectorizer(sparse=False)
dv_ng.fit(dicts_train_ng)

DictVectorizer(sparse=False)

In [155]:
dv_rt = DictVectorizer(sparse=False)
dv_rt.fit(dicts_train_rt)

DictVectorizer(sparse=False)

In [164]:
dv_nor = DictVectorizer(sparse=False)
dv_nor.fit(dicts_train_nor)

DictVectorizer(sparse=False)

In [157]:
dv_rpm = DictVectorizer(sparse=False)
dv_rpm.fit(dicts_train_rpm)

DictVectorizer(sparse=False)

In [158]:
dv_ng.get_feature_names()

['availability_365',
 'calculated_host_listings_count',
 'latitude',
 'longitude',
 'minimum_nights',
 'number_of_reviews',
 'reviews_per_month',
 'room_type=Entire home/apt',
 'room_type=Private room',
 'room_type=Shared room']

In [159]:
dv_rt.get_feature_names()

['availability_365',
 'calculated_host_listings_count',
 'latitude',
 'longitude',
 'minimum_nights',
 'neighbourhood_group=Bronx',
 'neighbourhood_group=Brooklyn',
 'neighbourhood_group=Manhattan',
 'neighbourhood_group=Queens',
 'neighbourhood_group=Staten Island',
 'number_of_reviews',
 'reviews_per_month']

In [165]:
dv_nor.get_feature_names()

['availability_365',
 'calculated_host_listings_count',
 'latitude',
 'longitude',
 'minimum_nights',
 'neighbourhood_group=Bronx',
 'neighbourhood_group=Brooklyn',
 'neighbourhood_group=Manhattan',
 'neighbourhood_group=Queens',
 'neighbourhood_group=Staten Island',
 'reviews_per_month',
 'room_type=Entire home/apt',
 'room_type=Private room',
 'room_type=Shared room']

In [166]:
dv_rpm.get_feature_names()

['availability_365',
 'calculated_host_listings_count',
 'latitude',
 'longitude',
 'minimum_nights',
 'neighbourhood_group=Bronx',
 'neighbourhood_group=Brooklyn',
 'neighbourhood_group=Manhattan',
 'neighbourhood_group=Queens',
 'neighbourhood_group=Staten Island',
 'number_of_reviews',
 'room_type=Entire home/apt',
 'room_type=Private room',
 'room_type=Shared room']

In [168]:
X_train_ng = dv_ng.transform(dicts_train_ng)

In [177]:
X_train_ng

array([[ 50.     ,  13.     ,  40.7276 , ...,   1.     ,   0.     ,
          0.     ],
       [  7.     ,   1.     ,  40.70847, ...,   0.     ,   1.     ,
          0.     ],
       [  0.     ,   1.     ,  40.83149, ...,   1.     ,   0.     ,
          0.     ],
       ...,
       [ 88.     ,   1.     ,  40.79994, ...,   0.     ,   1.     ,
          0.     ],
       [  0.     ,   1.     ,  40.69585, ...,   0.     ,   1.     ,
          0.     ],
       [281.     ,   2.     ,  40.64438, ...,   1.     ,   0.     ,
          0.     ]])

In [169]:
X_train_rt = dv_rt.transform(dicts_train_rt)

In [170]:
X_train_nor = dv_nor.transform(dicts_train_nor)

In [171]:
X_train_rpm = dv_rpm.transform(dicts_train_rpm)

In [173]:
model_ng = LogisticRegression(solver='lbfgs', C=1.0, random_state=42)
model_ng.fit(X_train_ng, y_train)

LogisticRegression(random_state=42)

In [174]:
model_rt = LogisticRegression(solver='lbfgs', C=1.0, random_state=42)
model_rt.fit(X_train_rt, y_train)

/Users/rashmi/mlzoomenv/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

In [175]:
model_nor = LogisticRegression(solver='lbfgs', C=1.0, random_state=42)
model_nor.fit(X_train_nor, y_train)

/Users/rashmi/mlzoomenv/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

In [176]:
model_rpm = LogisticRegression(solver='lbfgs', C=1.0, random_state=42)
model_rpm.fit(X_train_rpm, y_train)

/Users/rashmi/mlzoomenv/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:763: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(random_state=42)

In [178]:
X_val_ng = dv_ng.transform(dicts_val_ng)
X_val_rt = dv_rt.transform(dicts_val_rt)
X_val_nor = dv_nor.transform(dicts_val_nor)
X_val_rpm = dv_rpm.transform(dicts_val_rpm)

In [179]:
y_pred_ng = model_ng.predict_proba(X_val_ng)[:,1]
y_pred_rt = model_rt.predict_proba(X_val_rt)[:,1]
y_pred_nor = model_nor.predict_proba(X_val_nor)[:,1]
y_pred_rpm = model_rpm.predict_proba(X_val_rpm)[:,1]

In [180]:
ab_decision_ng = (y_pred_ng >= 0.5)
ab_decision_rt = (y_pred_rt >= 0.5)
ab_decision_nor = (y_pred_nor >= 0.5)
ab_decision_rpm = (y_pred_rpm >= 0.5)

In [203]:
acc_ng = (y_val == ab_decision_ng).mean().round(3)
acc_ng

0.751

In [204]:
acc_rt =(y_val == ab_decision_rt).mean().round(3)
acc_rt

0.717

In [209]:
acc_nor =(y_val == ab_decision_nor).mean().round(4)
acc_nor

0.787

In [208]:
acc_rpm =(y_val == ab_decision_rpm).mean().round(4)
acc_rpm

0.7853

In [200]:
diff_ng = acc_main - acc_ng
diff_rt = acc_main - acc_rt
diff_nor = acc_main - acc_nor
diff_rpm = acc_main - acc_rpm

In [202]:
print("Difference in accuracy after removing neighbourhood_group ",diff_ng ) 
print("Difference in accuracy after removing room_type ",diff_rt )
print("Difference in accuracy after removing number_of_reviews ",diff_nor )
print("Difference in accuracy after removing reviews_per_month ",diff_rpm )

Difference in accuracy after removing neighbourhood_group  0.039000000000000035
Difference in accuracy after removing room_type  0.07300000000000006
Difference in accuracy after removing number_of_reviews  0.0030000000000000027
Difference in accuracy after removing reviews_per_month  0.0050000000000000044


## Question 5 
### Which of following feature has the smallest difference?
### Difference in accuracy after removing number_of_reviews  0.00300 is the smallest 

In [210]:
dicts_full_train = df_full_train[categorical + numerical].to_dict(orient='records')

In [227]:
dicts_test = df_test[categorical + numerical].to_dict(orient='records')

In [211]:
dv = DictVectorizer(sparse=False)
X_full_train = dv.fit_transform(dicts_full_train)

In [228]:
X_test = dv.fit_transform(dicts_test)

In [231]:
X_test

array([[189.     ,   1.     ,  40.64354, ...,   1.     ,   0.     ,
          0.     ],
       [ 73.     ,   1.     ,  40.70666, ...,   0.     ,   1.     ,
          0.     ],
       [  0.     ,   1.     ,  40.76116, ...,   0.     ,   1.     ,
          0.     ],
       ...,
       [  5.     ,   4.     ,  40.75914, ...,   0.     ,   1.     ,
          0.     ],
       [  0.     ,   1.     ,  40.72179, ...,   1.     ,   0.     ,
          0.     ],
       [364.     ,   1.     ,  40.72978, ...,   0.     ,   1.     ,
          0.     ]])

In [212]:
y_train = np.log1p(df_full_train.price.values)
y_test = np.log1p(df_test.price.values)

In [213]:
y_train

array([5.69035945, 4.26267988, 4.07753744, ..., 4.87519732, 5.01727984,
       4.79579055])

In [214]:
y_test

array([4.49980967, 3.4339872 , 4.79579055, ..., 4.26267988, 6.2166061 ,
       4.83628191])

In [215]:
from sklearn.linear_model import Ridge

In [216]:
clf_zero = Ridge(alpha=0)
clf_zeroone = Ridge(alpha=0.01)
clf_pointone = Ridge(alpha=0.1)
clf_one = Ridge(alpha=1)
clf_ten = Ridge(alpha=10)

In [219]:
acc_zero  = clf_zero.fit(X_full_train, y_train)

In [220]:
acc_zeroone  = clf_zeroone.fit(X_full_train, y_train)
acc_pointone  = clf_pointone.fit(X_full_train, y_train)
acc_one  = clf_one.fit(X_full_train, y_train)
acc_ten  = clf_ten.fit(X_full_train, y_train)

In [224]:
def rmse(y, y_pred):
    se = (y - y_pred) ** 2
    mse = se.mean()
    return np.sqrt(mse)

In [247]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from math import sqrt

In [245]:
X_full_train.shape, y_train.shape, X_test.shape, y_test.shape

((39116, 15), (39116,), (9779, 15), (9779,))

In [248]:
pred_train_rr= acc_zero.predict(X_full_train)
print(np.sqrt(mean_squared_error(y_train,pred_train_rr)))
print(r2_score(y_train, pred_train_rr))

0.5002039586400198
0.48523992905743485


In [250]:
pred_test_rr = acc_zero.predict(X_test)
print("Score for 0 : ")
print(np.sqrt(mean_squared_error(y_test,pred_test_rr))) 
print(r2_score(y_test, pred_test_rr))

Score for 0 : 
0.48657189489884756
0.4995595534493944


In [251]:
pred_test_rr = acc_zeroone.predict(X_test)
print("Score for 0.01 : ")
print(np.sqrt(mean_squared_error(y_test,pred_test_rr))) 
print(r2_score(y_test, pred_test_rr))

Score for 0.01 : 
0.48645792270514343
0.49979396738147586


In [252]:
pred_test_rr = acc_pointone.predict(X_test)
print("Score for 0.1 : ")
print(np.sqrt(mean_squared_error(y_test,pred_test_rr))) 
print(r2_score(y_test, pred_test_rr))

Score for 0.1 : 
0.4864561589208569
0.49979759463842766


In [253]:
pred_test_rr = acc_one.predict(X_test)
print("Score for 1 : ")
print(np.sqrt(mean_squared_error(y_test,pred_test_rr))) 
print(r2_score(y_test, pred_test_rr))

Score for 1 : 
0.4864458492951737
0.4998187963225862


In [254]:
pred_test_rr = acc_ten.predict(X_test)
print("Score for 10 : ")
print(np.sqrt(mean_squared_error(y_test,pred_test_rr))) 
print(r2_score(y_test, pred_test_rr))

Score for 10 : 
0.4867771740597392
0.49913720408507267


## Question 6: 
### 0.01, 0.1 and 1 are very close to each other .. However Selected alpha = 1, least RMSE of all. 